### ISSUE TITLE
BatchNorm on test set: Using training set or test set statistics?

### ISSUE THREAD

**Issue by azinonos** [junyanz/pytorch-CycleGAN-and-pix2pix#693]
Hello,

First of all a big thank you to the contributors for making this great repository, it has been very helpful. I have a question regarding the batch normalization performed when training with Pix2Pix. From the original paper [(link)](https://arxiv.org/abs/1611.07004), it  is stated that:

> At inference time, we run the generator net in exactly the same manner as during the training phase. This differs from the usual protocol in that we apply dropout at test time, and we apply batch normalization [29] using the statistics of the test batch, rather than aggregated statistics of the training batch. This approach to batch normalization, when the batch size is set to 1, has been termed “instance normalisation” and has been demonstrated to be effective at image generation tasks [54].

So it seems like the test batch should be using the statistics calculated from the test set itself, and not the training set. I've been looking into the code and can't find exactly where that happens. Also the default for pix2pix is 'batch norm', not 'instance norm' like it was mentioned in some posts. Can the contributors specify if this normalisation approach is indeed happening, like in the paper? Thank you.

**Comment by junyanz** [comment]
Please see these [comments](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/blob/master/test.py#L54) for more details. There is an option called `--eval`. By default, the code will not use `--eval` so that the code only uses the statistics in the test batch. You can change the behavior by adding `--eval` in your command line.

**Comment by azinonos** [comment]
Great, thanks!
